feat(aws-sdk): add s3 and kinesis service extensions for aws-sdk instrumentation

jj22ee opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

- https://github.com/open-telemetry/opentelemetry-js-contrib/issues/2360

## Short description of the changes

- Add service extensions for S3 and Kinesis in AWS SDK instrumentation.
  - For S3, add bucket name to span if present
  - For Kinesis, add stream name to span if present
- This is a simple starting point for these new service extensions. The bucket/stream names are most important, but in the future, perhaps more data can be collected as span attributes.

Regarding some added constants to the `utils` file:
```
export const _AWS_S3_BUCKET = 'aws.s3.bucket';
export const _AWS_KINESIS_STREAM_NAME = 'aws.kinesis.stream.name';
```
- `_AWS_S3_BUCKET` should be added to [OTel JS Semantic Attributes in the future](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-semantic-conventions/src/trace/SemanticAttributes.ts).
  - See S3 specification: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/object-stores/s3.md
- There is no related specification for `_AWS_KINESIS_STREAM_NAME`, but [some AWS repos are already using `aws.kinesis.stream.name`](https://github.com/aws-observability/aws-otel-python-instrumentation/blob/2c00bd07eaaf703880a24a2fcfe874cdb4196678/aws-opentelemetry-distro/src/amazon/opentelemetry/distro/)


## Testing

1. Unit tests
2. Tested with:
  - Kinesis Client: observed `'aws.kinesis.stream.name': myNewKinesisStream`
```
    const client = new KinesisClient({ region: 'us-east-1' });
    const input = { // CreateStreamInput
      StreamName: "myNewKinesisStream", // required
      ShardCount: Number(3),
      StreamModeDetails: { // StreamModeDetails
        StreamMode: "ON_DEMAND", // required
      },
    };
    const command = new CreateStreamCommand(input);
    await client.send(command).then(data => {
      console.log(data);
    })
```
  - S3 Client: observed `'aws.s3.bucket': test-bucket-not-exist-or-accessible`
```
  const s3Client = new S3Client({ region: 'us-east-1' });
  const bucketName = 'test-bucket-not-exist-or-accessible';
  try {
    await s3Client.send(
      new ListObjectsCommand({
        Bucket: bucketName,
      }),
    ).then((data) => {
      console.log(data);
    });
  } catch (e) {
    if (e instanceof Error) {
      console.error('Exception thrown: ', e.message);
    }
  } finally {
    res.send('done aws sdk s3 request');
  }
```

jj22ee commented:
Pinging @blumamir for review!

Unrelated heads-up, I'm looking to take over component ownership of AWS JS components from @carolabadeer. I think I just need to make a PR like https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1464.

trentm commented:
> I think I just need to make a PR like #1464.

@jj22ee Yup that's correct. Great to have you helping.

blumamir commented on plugins/node/opentelemetry-instrumentation-aws-sdk/package.json:
```diff
@@ -52,6 +52,7 @@
   },
   "devDependencies": {
     "@aws-sdk/client-dynamodb": "3.85.0",
+    "@aws-sdk/client-kinesis": "3.85.0",
```
is it used anywhere? 
seems the tests for kinesis uses v2 (`aws-sdk`) and this dependency is redundant.

blumamir commented on plugins/node/opentelemetry-instrumentation-aws-sdk/test/s3.test.ts:
```diff
@@ -0,0 +1,75 @@
+import { AwsInstrumentation } from '../src';
+registerInstrumentationTesting(new AwsInstrumentation());
+
+import * as AWS from 'aws-sdk';
```
awesome to have tests for that.

Please note that those tests are for v2 which [will soon be end of life](https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-javascript-v2/).
Consider also adding assertions on these new attributes to v3 tests in `test/aws-sdk-v3.test.ts` [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/60a99c98fd3a5594c7c2234184f06166b375e707/plugins/node/opentelemetry-instrumentation-aws-sdk/test/aws-sdk-v3.test.ts#L66)

blumamir reviewed (commented):
Thank you for the contribution and the info in the PR.

Left few comments. Some of them I wrote on s3 implementation but they also apply to kinesis, didn't want to duplicate the same comments twice :)

jj22ee commented:
Hi, looking for another round of review! I removed some of the aws-sdk v2 tests I had previously given their end-of-life and soon-to-be-deprecated statuses, and removed the latency with the Kinesis Client test (enforced HTTP instead of HTTP2 which isn't supported by `nock`)

blumamir reviewed (approved):
LGTM 💯 

Thank you for the contribution, the code and tests looks great and the description on the PR is very helpful.
